### DISCUSSION TITLE
ASGI applications can't send 101 Continue

### DISCUSSION THREAD

**Discussion by florimondmanca** [frankie567/httpx-ws#14] [Q&A] [answered]
Hey there,

Thanks a lot for hopping onto this HTTPX extension.

I'm trying to integrate `httpx-ws` on https://github.com/florimondmanca/arel/pull/31, where I'd like to test against an ASGI app using `AsyncClient(app=...)`.

The confusing bit I'm having a hard time with, is `httpx-ws` seems to expect that the ASGI application under test returns the `101 Continue` upgrade response. But this is generally done by ASGI _servers_. When a WebSocket request is received (on `ws://`), servers run the app with `scope = {"type": "websocket", ...}`. Applications catch this scope type and expect a `websocket.connect` message, then send `websocket.accept`, at which point the _server_ sends the 101 Continue response and upgrades the connection.

This might also be why browsers require `new WebSocket("ws://somewhere.org")` -- how would they know if they should initiate a WebSocket handshake or a regular HTTP request if both use `http://`?

I'm unsure what should change at this point, as I haven't dived into the code yet, but mostly like that would be the ASGI implementation which should run the app as an ASGI _server_ would? And use `ws://` for URLs rather than `http://`?

**Comment by florimondmanca** [comment, marked as answer]
Oooh, answering the question for myself -- I forgot to use `ASGIWebSocketTransport`...

That is correctly covered in the docs: https://frankie567.github.io/httpx-ws/usage/asgi/

So, fine!

**Reply by frankie567** [reply]
Exactly! The transport implements the logic to "emulate" an ASGI server by executing your app and mapping ASGI messages to a HTTPX network stream.

Regarding the scheme, HTTPX currently only supports `http`/`https`, that's why I don't show examples with `ws`/`wss`. But Tom has recently implemented a change in httpcore to support them: https://github.com/encode/httpcore/pull/625
